bin/validator_tests.py

```diff
@@ -1,20 +1,25 @@
 from collections.abc import Callable
-from typing import Optional, TypeVar
+from typing import Final, Iterable, Optional, Sequence, TypeVar
 
-_GENERATOR_NAMES: set[str] = set()
-GENERATORS: list[tuple[str, str | Callable[[str], Optional[str]], bool]] = []
+# we use an underspecified type so mypy warns if you add something manually
+_GENERATOR_NAMES: Final[Iterable[str]] = set()
+GENERATORS: Final[Sequence[tuple[str, str | Callable[[str], Optional[str]], bool]]] = []
 
 
 T = TypeVar("T", bound=str | Callable[[str], Optional[str]])
 
 
+# returns a function that can be called to register a new generator for invalid tests
+# can be used on its own or as decorator for a function
 def register(name: Optional[str] = None, only_whitespace_change: bool = False) -> Callable[[T], T]:
     def decorator(func: T) -> T:
         nonlocal name
         if not isinstance(func, str) and not name:
             assert hasattr(func, "__name__")
             name = func.__name__
         assert name
+        assert isinstance(_GENERATOR_NAMES, set)
+        assert isinstance(GENERATORS, list)
```

Comment:
If I can be honest, I think type-juggling immutable types back to mutable types is slightly uglier than writing one big function that returns the list of generators, instead of storing them in global variables :innocent: